tempesta_fw/sock_srv.c

```diff
@@ -298,13 +308,34 @@ tfw_sock_srv_connect_complete(struct sock *sk)
 	return 0;
 }
 
-static int
-tfw_sock_srv_do_failover(struct sock *sk, const char *msg)
+/**
+ * The hook is executed when we intentionally close a server connection during
+ * shutdown process. Now @sk is closed (but still alive) and we release all
+ * associated resources before SS put()'s the socket.
+ */
+static void
+tfw_sock_srv_connect_drop(struct sock *sk)
+{
+	TfwConnection *conn = sk->sk_user_data;
+
+	tfw_connection_drop(conn);
+	tfw_connection_put(conn);
+}
```

Comment:
Generally I explained the design decisions above.

We don't need to unlink Tempesta from sk, since it must be guaranteed that TfwConnection live longer than sk. BTW I strongly dislike current set of link/unlink calls: it's really hard to manage the links. I thought about aggregation of some of the calls, but didn't find a quick solution, so I left things as they are. Please think about the code simplification as well.

> Stats are not collected;

fixed

`tfw_connection_get_if_nfo()` doesn't return a connection if it's reference count is zero. Also `tfw_connection_put()` don't destruct a connection if its reference counter isn't zero. However, freed connections stay in schedulers' data structures and we could get a crash. This is why we must stop sock_srv module _after_ sock_clnt to guarantee that there is no scheduling requests when all server connections are dead.